_posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md

```diff
@@ -0,0 +1,201 @@
+
+### Step 1: Setting Up the environment
+As you may have known, a typical Metal3 workflow requires existence of several components: bootstrap kubernetes cluster, special networks, baremetal nodes, etc. As we are working on simulating environment, we will start with a newly spawn Ubuntu VM, create a cluster with minikube, add networks with libvirt, and so on (If you're familiar with the Metal3's dev-env, this step is similar to what script [01](), [02] and a part of [03] do). We will not go into details on this part, but you can find the related setup from [this script]() if you are interested.
+
+**Note**: If you intend to follow along, note that going to 1000 nodes requires a large environment and will take a long time. In our setup, we had a VM with 24 cores and 32GB of RAM, of which we assigned 14 cores and 20GB of RAM to the minikube VM, and the whole process took roughly 48 hours. If your environment is less powerful, consider reducing the number of nodes you want to provision. Something like 100 nodes will require very minimum resource and time, while still be an impressive outcome.
+### Step 2: Install BMO and Ironics
+
+In Metal3’s typical workflow, we usually rely on **Kustomize** to install Ironic and BMO. Kustomize helps us define configurations for Kubernetes resources, making it easier to customize and deploy services. However, the current Kustomize overlay we use for Metal3 is designed to configure **only a single Ironic instance**. This setup works well for smaller environments, but it becomes a bottleneck when you’re trying to scale up and handle thousands of nodes.
+
+That’s where Ironic’s **special mode** comes into play. Ironic has the ability to run **multiple Ironic conductors** while sharing the same database. The best part? Workload balancing between conductors happens automatically. This means that no matter which Ironic conductor receives a request, the load is evenly distributed across all conductors, ensuring efficient provisioning. To achieve this requires separation of **ironic conductor** from the database, from which the conductor part can be scaled up. Each **conductor** will have its own `PROVISIONING_IP`, hence the need of having a specialized `configMap`.
+
+We used [Helm]() for this purpose. In our Helm chart, **Ironic conductor** container and **HTTP server (httpd)** container are separated into a new pod, while the rest of the ironic package (mostly `mariadb` - ironic database) stays in another pod. A list of `PROVISIONING_IP`s is provided by the chart's `values.yaml`, and for each ip, an  **ironic conductor** pod is created, along with a configmap whose values rendered with the ip's value. This way, we can dynamically scale up/down ironic (or more specifically **ironic conductors** by simply adding/removing ips.
```

Comment:
Missing link for Helm, or the link syntax is not needed.

Reply:
Yeah. As this a WIP, I was writing down markdown link syntax just as placeholders. No proper links were added. Probably I could add the TODO to remember adding them later.